### ISSUE TITLE
TimeoutException when Maxwell tries to connect to RabbitMQ

### ISSUE THREAD

**Issue by tadavid-cae** [zendesk/maxwell#1890]
We are having trouble starting maxwell on specific servers. Other servers working just fine, but there is no obvious difference between them. The error says TimeoutException.  Based on the trace it is a problem during establishing the AMQP connection towards RabbitMQ. **Please help us since this is a critical problem regarding our operation and we could not find any solution to this. 🙏🏻** 

---
**Maxwell version:**
```
1.37.5
```
**Java version:**
```
openjdk 11.0.14.1 2022-02-08 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.14.1+1-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.14.1+1-LTS, mixed mode, sharing)
```
**RabbitMQ version:**
```
rabbitmq-server-3.8.16-1.el7.noarch
```
---

We eliminated the network and rabbitmq credential issue since other services can connect to RabbitMQ. If we change the maxwell configuration to use wrong rabbitmq credentials the exact same error happens. 
I have attached the TCP dump [maxwell-tcpdump.zip](https://github.com/zendesk/maxwell/files/9110072/maxwell-tcpdump.zip) from the communication between maxwell. It will shows the maxwell instantly terminates the connection: 

![MicrosoftTeams-image (35)](https://user-images.githubusercontent.com/81415962/178954854-70fc13af-8895-4921-9b26-64756040c196.png)

---
**config.properties of maxwell**

```
log_level=info
producer=rabbitmq
user=x
password=x
host=localhost
port=3306
schema_database=maxwell
filter=include: x.users, include: x.new_submits
rabbitmq_host=localhost
rabbitmq_port=5672
rabbitmq_user=x
rabbitmq_pass=x
rabbitmq_virtual_host=/
rabbitmq_exchange=maxwell
rabbitmq_exchange_type=topic
rabbitmq_exchange_durable=true
rabbitmq_exchange_autodelete=false
rabbitmq_routing_key_template=cdc.%db%.%table%
rabbitmq_message_persistent=true
rabbitmq_declare_exchange=true

```

---
**Maxwell service unit:**

```
[Unit]
Description=Maxwell Change Detection Capture
After=sockets.target aa_lionis_imager_first.service

[Install]
WantedBy=x.target

[Service]
User=x
Group=x

Type=simple
Environment='JAVA_OPTS=-Xms64m -Xmx1024m'
ExecStart=/cae/maxwell/bin/maxwell --config /cae/maxwell/config.properties

RestartSec=2s
Restart=on-failure
StartLimitBurst=3
StartLimitInterval=30

```
---
**RabbitMQ log from the failing connection:**

```
2022-07-14 03:59:43.746 [info] <0.1599.0> accepting AMQP connection <0.1599.0> (127.0.0.1:58622 -> 127.0.0.1:5672)
2022-07-14 03:59:51.748 [warning] <0.1599.0> closing AMQP connection <0.1599.0> (127.0.0.1:58622 -> 127.0.0.1:5672):
client unexpectedly closed TCP connection
2022-07-14 03:59:51.749 [info] <0.1603.0> Closing all channels from connection '127.0.0.1:58622 -> 127.0.0.1:5672' because it has been closed

```
---

**Maxwell Service log:**

```
Jul 05 14:04:30 server systemd[1]: Started Maxwell Change Detection Capture.
Jul 05 14:04:37 server maxwell[54554]: 6007 [INFO] MLog: MLog clients using slf4j logging.
Jul 05 14:04:37 server maxwell[54554]: 6162 [INFO] C3P0Registry: Initializing c3p0-0.9.5.5 [built 11-December-2019 22:18:33 -0800; debug? true; trace: 10]
Jul 05 14:04:37 server maxwell[54554]: 6419 [INFO] AbstractPoolBackedDataSource: Initializing c3p0 pool... com.mchange.v2.c3p0.ComboPooledDataSource [ acquireIncrement -> 3, acquireRetryAttempts -> 1, acquireRetryDelay -
Jul 05 14:04:38 server maxwell[54554]: 7104 [INFO] AbstractPoolBackedDataSource: Initializing c3p0 pool... com.mchange.v2.c3p0.ComboPooledDataSource [ acquireIncrement -> 3, acquireRetryAttempts -> 1, acquireRetryDelay -
Jul 05 14:04:38 server maxwell[54554]: 7144 [INFO] AbstractPoolBackedDataSource: Initializing c3p0 pool... com.mchange.v2.c3p0.ComboPooledDataSource [ acquireIncrement -> 3, acquireRetryAttempts -> 30, acquireRetryDelay
Jul 05 14:04:38 server maxwell[54554]: 7223 [INFO] Maxwell: Starting Maxwell. maxMemory: 134217728 bufferMemoryUsage: 0.25
Jul 05 14:04:38 server maxwell[54554]: 7236 [INFO] AbstractPoolBackedDataSource: Initializing c3p0 pool... com.mchange.v2.c3p0.ComboPooledDataSource [ acquireIncrement -> 3, acquireRetryAttempts -> 30, acquireRetryDelay
Jul 05 14:04:38 server maxwell[54554]: 7301 [INFO] SchemaStoreSchema: Creating maxwell database
Jul 05 14:04:38 server maxwell[54554]: 7459 [INFO] AbstractPoolBackedDataSource: Initializing c3p0 pool... com.mchange.v2.c3p0.ComboPooledDataSource [ acquireIncrement -> 3, acquireRetryAttempts -> 30, acquireRetryDelay
Jul 05 14:04:43 server maxwell[54554]: 12672 [WARN] ForgivingExceptionHandler: An unexpected connection driver error occured (Exception message: Socket closed)
Jul 05 14:04:43 server maxwell[54554]: 12689 [INFO] TaskManager: Stopping 0 tasks
Jul 05 14:04:43 server maxwell[54554]: 12690 [ERROR] TaskManager: cause:
Jul 05 14:04:43 server maxwell[54554]: java.lang.RuntimeException: java.util.concurrent.TimeoutException
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.producer.RabbitmqProducer.<init>(RabbitmqProducer.java:65) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.MaxwellContext.getProducer(MaxwellContext.java:548) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:257) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.start(Maxwell.java:227) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.main(Maxwell.java:336) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: Caused by: java.util.concurrent.TimeoutException
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingCell.get(BlockingCell.java:77) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingCell.uninterruptibleGet(BlockingCell.java:120) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:36) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:502) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.AMQConnection.start(AMQConnection.java:317) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.recovery.RecoveryAwareAMQConnectionFactory.newConnection(RecoveryAwareAMQConnectionFactory.java:64) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.init(AutorecoveringConnection.java:156) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1106) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1063) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1021) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1182) ~[amqp-client-5.7.3.jar:5.7.3]
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.producer.RabbitmqProducer.<init>(RabbitmqProducer.java:60) ~[maxwell-1.37.5.jar:1.37.5]
Jul 05 14:04:43 server maxwell[54554]: ... 4 more
Jul 05 14:04:43 server maxwell[54554]: 12711 [INFO] TaskManager: Stopped all tasks
Jul 05 14:04:43 server maxwell[54554]: java.lang.RuntimeException: java.util.concurrent.TimeoutException
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.producer.RabbitmqProducer.<init>(RabbitmqProducer.java:65)
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.MaxwellContext.getProducer(MaxwellContext.java:548)
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:257)
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.start(Maxwell.java:227)
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.Maxwell.main(Maxwell.java:336)
Jul 05 14:04:43 server maxwell[54554]: Caused by: java.util.concurrent.TimeoutException
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingCell.get(BlockingCell.java:7
[maxwell.zip](https://github.com/zendesk/maxwell/files/9109903/maxwell.zip)
7)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingCell.uninterruptibleGet(BlockingCell.java:120)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:36)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:502)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.AMQConnection.start(AMQConnection.java:317)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.recovery.RecoveryAwareAMQConnectionFactory.newConnection(RecoveryAwareAMQConnectionFactory.java:64)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.impl.recovery.AutorecoveringConnection.init(AutorecoveringConnection.java:156)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1106)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1063)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1021)
Jul 05 14:04:43 server maxwell[54554]: at com.rabbitmq.client.ConnectionFactory.newConnection(ConnectionFactory.java:1182)
Jul 05 14:04:43 server maxwell[54554]: at com.zendesk.maxwell.producer.RabbitmqProducer.<init>(RabbitmqProducer.java:60)
Jul 05 14:04:43 server maxwell[54554]: ... 4 more
Jul 05 14:04:43 server systemd[1]: maxwell.service: main process exited, code=exited, status=1/FAILURE
Jul 05 14:04:43 server systemd[1]: Unit maxwell.service entered failed state.
Jul 05 14:04:43 server systemd[1]: maxwell.service failed.
Jul 05 14:04:45 server systemd[1]: maxwell.service holdoff time over, scheduling restart.
Jul 05 14:04:45 server systemd[1]: Stopped Maxwell Change Detection Capture.
```

**Comment by osheroff** [comment]
that's a of emoji's on this.  you all must work together.

The stack trace indicates that the server simply isn't responding to the very first packet that the client is sending.  It's strange, because some network activity seems to be taking place (the heartbeats are getting ack'ed), but there's nothing coming from the server.

can you run a tcpdump on the server side and see if it's receiving packets?  are there any logs on the server of interest?

**Comment by tadavid-cae** [comment]
I will send you the tcpdump. 

The server of interest is a rabbitmq instance. Which is only logging this regarding the connection attempt:
```
2022-07-14 03:59:43.746 [info] <0.1599.0> accepting AMQP connection <0.1599.0> (127.0.0.1:58622 -> 127.0.0.1:5672)
2022-07-14 03:59:51.748 [warning] <0.1599.0> closing AMQP connection <0.1599.0> (127.0.0.1:58622 -> 127.0.0.1:5672):
client unexpectedly closed TCP connection
2022-07-14 03:59:51.749 [info] <0.1603.0> Closing all channels from connection '127.0.0.1:58622 -> 127.0.0.1:5672' because it has been closed
```

**Comment by osheroff** [comment]
Also, please try to capture a trace from the *beginning* of the connection; this is somewhere in the middle.

**Comment by tadavid-cae** [comment]
@osheroff  do you have any other suggestion which may help us? 
The rabbitmq and maxwell are running on the same server. We have multiple servers which are running this identical setup, however only couple of them having this connection problem. We could not find any difference between the server setups. Is there any deep/hack/notwellknown JVM setting that may cause this problem?

**Comment by wanghangyu817** [comment]
A piece of code that may raise an exception：
`
this.channel = factory.newConnection().createChannel();
			if(context.getConfig().rabbitmqDeclareExchange) {
				this.channel.exchangeDeclare(exchangeName, context.getConfig().rabbitmqExchangeType, context.getConfig().rabbitMqExchangeDurable, context.getConfig().rabbitMqExchangeAutoDelete, null);
			}
`

- You are advised to check the localhost mapping.  Is there any special configuration?
- Check the local loopback address and port
- Try to test the connection using the native IP, which may work
 To sum up, is my understanding, hope to correct

**Comment by 95gabor** [comment]
The problem was that the dns name resolution was slow due to a slow network. The error was solved by setting the handshake timeout higher. (with a custom maxwell build)

**Comment by tadavid-cae** [comment]
@wanghangyu817 you could easily repro the issue if you add a delay to your network interface 
```
tc qdisc add dev eth0 root netem delay 2000ms
```

**Comment by osheroff** [comment]
latest release has the configurable timeout.  LMK how it goes.

**Comment by 95gabor** [comment]
> latest release has the configurable timeout. LMK how it goes.

Thanks! Looks like it solved our problem.
